sympy/combinatorics/coset_table.py

```diff
@@ -1014,3 +1016,169 @@ def coset_enumeration_c(fp_grp, Y, max_cosets=None, draft=None,
+        s[rho] = lambda_
+        lambda_ = rho
+        rho = p[lambda_]
+    rho = s[lamda]
+    while rho != k:
+        mu = rho
+        rho = s[mu]
+        p[rho] = lambda_
+        p_p[rho] = p_p[rho]*p_p[mu]
+    return lambda_
+
+def modified_merge(C, k, lamda, w, p_p, q, l):
+    '''
+    Input:κ,λ∈Ω,w∈(Y∪Y–1)*withτ(κ)= φ(w)τ(λ),p_p,q,l GP
+    q is a queue of length l of elements to be deleted from Ω *
+    '''
+    phi = modified_rep(C, k, p_p)
+    psi = modified_rep(C, lamda, p_p)
+    if phi > psi:
+        C.p[phi] = psi
+        p_p[phi] =  p_p[k]**-1*w*p_p[lamda]
+        l += 1
+        q[l] = psi
+    elif phi < psi:
+        C.p[psi] = phi
+        p_p[psi] = p_p[lamda]**-1*w**-1*p_p[k]
+        l += 1
+        q[l] = psi
+
+def modified_coincidence(C, alpha, beta, w):
+    '''
+    Input: A coincident pair α, ß∈Ω, w∈ (Y∪Y–1)*
+    '''
+    l = 0
+    i = 0
+    q = []
+    p_p = []
+    table = C.table
+    modified_merge(C, alpha, beta, w, p_p, q, l)
+    while len(q):
+        gamma = q.pop(0)
+        for x in C.A_dict:
+            delta = C.table[gamma][C.A_dict[x]]
+        if delta:
+            table[delta][C.A_dict_inv[x]] = None
+            mu = modified_rep(C, gamma, p_p)
+            nu = modified_rep(C, delta, p_p)
+        if table[mu][C.A_dict[x]]:
+            v = p_p[delta]**-1*C.P[gamma][C.A_dict[x]]**-1*p_p[gamma]*C.P[mu][C.A_dict[x]]
+            modified_merge(C, nu, table[mu][C.A_dict[x]], v, p_p, q, l)
+        elif table[nu][C.A_dict_inv[x]]:
+            v = p_p[gamma]**-1*C.P[gamma][C.A_dict[x]]*p_p[delta]*C.P[mu][C.A_dict_inv[x]]
+            modified_merge(C, mu, table[nu][C.A_dict_inv[x]], v, p_p, q, l)
+        else:
+            table[mu][C.A_dict[x]] = nu
+            table[nu][C.A_dict_inv[x]] = mu
+            v = p_p[gamma]**-1*C.P[gamma][C.A_dict[x]]*p_p[delta]
+            C.P[mu][C.A_dict[x]] = v
+            C.P[nu][C.A_dict_inv[x]] = v**-1
+
+def modifed_coset_enumeration_r(fp_grp, Y, max_cosets=None, draft=None, incomplete=False):
+    C = CosetTable(fp_grp, Y, max_cosets=max_cosets)
+    if draft:
+        C.table = draft.table[:]
+        C.p = draft.p[:]
+    R = fp_grp.relators
+    A_dict = C.A_dict
+    A_dict_inv = C.A_dict_inv
+    p = C.p
+    for w in Y:
+        C.modifed_scan_and_fill(0, w)
+    alpha = 0
+    while alpha < C.n:
+        if p[alpha] == alpha:
+            try:
+                for w in R:
+                    C.modifed_scan_and_fill(alpha, w)
+                    # if α was eliminated during the scan then break
+                    if p[alpha] < alpha:
+                        break
+                if p[alpha] == alpha:
+                    for x in A_dict:
+                        if C.table[alpha][A_dict[x]] is None:
+                            C.modifed_define(alpha, x)
+            except ValueError as e:
+                if incomplete:
+                    return C
+                raise e
+        alpha += 1
+    return C
```

Comment:
Is the above code essentially the same as `coset_enumeration_r` but with all relevant methods replaced by modified versions? If so, perhaps, we should just have a keyword `modified` in `coset_enumeration_r` and pick the right methods at the start, e.g. `_define = C.define` or `_define = C.modified_define` and use `_define` where necessary.

Reply:
What about this?

Reply:
I push it with the unit tests.